### DISCUSSION TITLE
API restructuring proposal

### DISCUSSION THREAD

**Discussion by nyurik** [chridou/condow#38] [Ideas]
I did a first pass at http implementation in https://github.com/chridou/condow/pull/36, and that gave me a few API restructuring ideas I would like to discuss:

### API restructuring ideas
* Core should not know `what` it is downloading. Location should be part of the source's configuration, and should be source specific - i.e. a local file, URL, bucket+path, etc.
* Core should have a user facing API, e.g.  `condow::new(source).some_option(...).other_option(..).run().await?`.  Each source instance would follow the same pattern with the options specific to that source.
* Move source-specific options like retries into the source (i.e. why retry an FS source?)
* It should be possible to add more than one source, maybe even of different types. Makes it possible to download from multiple mirrors.
* If possible, get rid of the additional config object - make it part of the main `condow` object
* Downloading just a part of a file via this system is too confusing - `download(url, ..)` (the two dots). Move `range` param to the options.
* Saving to a file is probably the most common usage pattern - need that as a `condow` option, e.g. `.save_as(file)`, optimized internally (i.e. possibly with memmaped files, etc)
* logging situation is very confusing - I couldn't figure out how to easily set up a logger to see what condow was doing
* sources don't have access to the central logger, need to pass it in somehow?

P.S. This was moved from #35

**Comment by chridou** [comment]
> Core should not know what it is downloading. Location should be part of the source's configuration, and should be source specific - i.e. a local file, URL, bucket+path, etc.



I am not sure about this one. The way I am using it now is that I am downloading multiple BLOBs from the same source (even multiple files concurrently).

I am using it like 

```
let condow = condow::new(s3_client, config);
condow.download(bucket_a/blob_a);
condow.download(bucket_a/blob_b);
condow.download(bucket_c/blob_c);
```

Also the condow instances has (will have) some shared state like metrics collectors and later on a "resource manager".

Would something like that still be possible since the location would be part of the configuration

**Reply by nyurik** [reply]
Thanks, this is a valid use case too. I wonder how we can reconcile them. We have two basic features:
* A downloader that knows how to partition a single file and write it to an output, using one or more sources.
* A download manager that coordinates/balances multiple downloaders, keeps overall statistics, limits concurrent downloads per HTTP client and for the whole system.

Do you think it would simplify things to separate these two [concerns](https://en.wikipedia.org/wiki/Separation_of_concerns)? 

Some possible usage with the manager:

```
let condow = condow_manager::new(global_config);   // or could be:   new().option_a(...).option_b(...);
let s3 = s3_cilent::new(creds);
condow.download(s3.source(bucket_a/blob_a));
condow.download(s3.source(bucket_a/blob_b));
condow.download_from_multiple(vec![
    s3.source(bucket_c/blob_c),
    http_client(blob_c),
]);
```

or the simpler usage without the manager:
```
let condow = condow::new(source).some_option(...).other_option(..).run().await?
```

P.S. (edited):
Updated first example - most likely s3 client sources will want shared state, which is different from global shared state/config.

**Reply by nyurik** [reply]
P.P.S. Technically we could still keep both manager and partitioner together, and just use the above first (altered) proposal with `condow_manager`. It seems very similar to your example, just adding the s3 client as a separate entity - and would solve most of my concerns.

**Reply by nyurik** [reply]
@chridou hi, any updates on this? Thx!!

**Reply by chridou** [reply]
Hi! Yes. Currently I am switching the logging to `tracing`.
Next would be the API and then the hard part would be to change the model of execution...

**Reply by nyurik** [reply]
awesome, thanks @chridou for all the hard work!  Let me know how it progresses.  Are you ok with my latest proposal, or do you think we should revise it further? Thx!

**Comment by chridou** [comment]
> Core should have a user facing API, e.g. condow::new(source).some_option(...).other_option(..).run().await?. Each source instance would follow the same pattern with the options specific to that source.

Is this just adding parameters for a download. A "fluent API"...

`let stream = condow.download(range(..)).await`

Yes!!!

EDITED

**Comment by chridou** [comment]
> Move source-specific options like retries into the source (i.e. why retry an FS source?)

It is actually there because I had the assumption that all network stuff works roughly the same. I wanted to have the internal client interface to be easily implementable (and implementations simple).   I see that it might get a little more complicated now...

FS support is actually not the target. The FS crate is just there for loading files from the FS while testing. The FS crate says that it is mostly for testing (replacing a network based client for testing to not have flaky tests). But yes. For the FS the retries make no sense.

Maybe we could have the retry mechanism as part of a toolbox so that it can be used within clients where necessary.

**Reply by nyurik** [reply]
heh, yeah, i didn't think FS was a big target here :)  (and now someone will come in and tell us that it works better over NFS to do it in multipart too...)

Also, it definitely should go beyond retries - i'm sure there will be other client-specific options.

The issue with retries and timeouts is that many client-specific libraries already implement that, e.g. for the `reqwest` there is [reqwest-retry](https://lib.rs/crates/reqwest-retry) middleware, plus its own custom timeout system. Maybe retries are worth making globally, vs timeouts should be a client-specific param?  Or maybe users will want per-client retry settings and having a global one will make it more restrictive - i don't know tbh, and i don't have a strong opinion on this one.

**Reply by chridou** [reply]
The thing that I was kinda "proud of" was that the retries in condow can also continue on streams aborted by IO errors. Maybe wee only keep that feature? I don't know... Maybe we have a look at it again later?

**Reply by nyurik** [reply]
Sure, i guess retries could be implemented in the core too -- i was just contemplating here. The very first one is far more important to me.

**Comment by chridou** [comment]
> It should be possible to add more than one source, maybe even of different types. Makes it possible to download from multiple mirrors.

Yes! For HTTP download (or other mirrored sources) this perfectly makes sense. I think this is where it gets a bit complicated and the "simple dumb client implementations) are not sufficient anymore.

**Reply by nyurik** [reply]
agree, see my altered proposal in https://github.com/chridou/condow/discussions/38#discussioncomment-2200509

**Comment by chridou** [comment]
> Downloading just a part of a file via this system is too confusing - download(url, ..) (the two dots). Move range param to the options.

Yes. Also the API for collecting data on the downloads (the reporting stuff) which makes the API ugly.

**Comment by chridou** [comment]
> Saving to a file is probably the most common usage pattern - need that as a condow option, e.g. .save_as(file), optimized internally (i.e. possibly with memmaped files, etc)

Memmapped files would be cool! Maybe add something like this as a feature?
I'd wish we could keep the number of required dependencies to a minimum.

Actually my use case is downloading into memory. I admit that I did not really have files in mind and thought that the streams implement `AsyncRead` would be enough.

**Reply by nyurik** [reply]
heh, memmap could always be hidden behind a feature flag, plus it is definitely not the top priority. On the other hand, saving files to disk is something most users would expect - basic thinking is that if you download via multiple parts, the file is substantially large to make it worth while, and therefor could easily be bigger than available RAM.

**Comment by chridou** [comment]
> logging situation is very confusing - I couldn't figure out how to easily set up a logger to see what condow was doing

True. Currently the `Reporter` (metrics collector interface) is misused.

Actually I did not yet  think about a proper solution since as far as I know there is no logging facade yet which supports contextual/semantic logging. 
A trait to be used would be fine for me where users could plug their logger.

**Comment by chridou** [comment]
> sources don't have access to the central logger, need to pass it in somehow?

Yes. That would be the solution. Actually I was already thinking about passing the metrics `Reporter`, too, but refrained from it. The `CondowClient` trait is used only internally anyways for now so we could pass what we want via it's methods.
